File: app/src/main/java/sembako/sayunara/android/ui/component/product/listproduct/adapter/ProductAdapter.java
package sembako.sayunara.android.ui.component.product.listproduct.adapter;

import android.content.Context;
import android.graphics.Paint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.desmond.squarecamera.SquareImageView;
import com.google.android.material.button.MaterialButton;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;

import sembako.sayunara.android.R;
import sembako.sayunara.android.ui.component.product.listproduct.model.Product;
import sembako.sayunara.android.ui.util.HFRecyclerViewAdapter;


public class ProductAdapter extends HFRecyclerViewAdapter<Product, ProductAdapter.ItemViewHolder> {
    private OnItemClickListener onItemClickListener;
    int minteger = 0;
    protected int total = 0;
    protected boolean isHome = false;
    int price = 0;
    Context context;
    interface OnItemSelected {
        void onSelect(Product model);
    }

    private OnItemSelected listener;


    public ProductAdapter(Context context1, boolean isHome) {
        super(context1);
        context = context1;
        this.isHome = isHome;
    }

    @Override
    public void footerOnVisibleItem() {

    }


    @Override
    public ItemViewHolder onCreateDataItemViewHolder(ViewGroup parent, int viewType) {
        View v;
        if(isHome){
            v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_home, parent, false);
        }else {
            v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_product, parent, false);
        }

        return new ItemViewHolder(v);
    }

    @Override
    public void onBindDataItemViewHolder(ItemViewHolder holder, final int position) {
        Product product = getData().get(position);

        holder.name.setText(product.getName());

        if(!product.getImages().get(0).isEmpty()){
            Picasso.get()
                    .load(product.getImages().get(0))
                    .into(holder.image, new Callback() {
                        @Override
                        public void onSuccess() {
                            holder.progressBar.setVisibility(View.GONE);
                        }

                        @Override
                        public void onError(Exception e) {
                            holder.progressBar.setVisibility(View.VISIBLE);
                        }
                    });
        }

        double harga = Double.parseDouble(String.valueOf(product.getPrice()));
        DecimalFormat df = (DecimalFormat) DecimalFormat.getCurrencyInstance();
        DecimalFormatSymbols dfs = new DecimalFormatSymbols();
        dfs.setCurrencySymbol("");
        dfs.setMonetaryDecimalSeparator(',');
        dfs.setGroupingSeparator('.');
        df.setDecimalFormatSymbols(dfs);
        String k = df.format(harga);


        if(product.getDiscount()==0){
            holder.button_discount.setVisibility(View.GONE);

            holder.price.setText("Rp"+k+" /");
            holder.tv_product_unit.setText(product.getWeight()+" "+product.getUnit());

        }else {
            holder.button_discount.setVisibility(View.VISIBLE);
            holder.textview_price_discount.setVisibility(View.VISIBLE);
            holder.button_discount.setText(String.valueOf(product.getDiscount())+ " %");
            int price = Integer.parseInt(String.valueOf(product.getPrice()));
            int discount = Integer.parseInt(String.valueOf(product.getDiscount()));

            int total = (discount*price)/100;
            int total2 = price-total;
            double amount = Double.parseDouble(String.valueOf(total2));
            holder.textview_price_discount.setPaintFlags(holder.textview_price_discount.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

            holder.textview_price_discount.setText("Rp"+k+" / "+product.getWeight()+" "+product.getUnit());
            holder.tv_product_unit.setText(product.getWeight()+" "+product.getUnit());


            holder.price.setText("Rp"+df.format(amount)+" /");
        }

        holder.btn_buy.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClickListener.OnActionClickQuestion(v,position);

               /* minteger = 1;
                holder.tv_integer_number.setText(String.valueOf(minteger));
                holder.btn_buy.setVisibility(View.GONE);
                holder.linear_quantity.setVisibility(View.VISIBLE);*/
            }
        });

        holder.btn_decrease.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

            }
        });

        holder.btn_increase.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                minteger = minteger + 1;
                holder.tv_integer_number.setText(String.valueOf(minteger));
            }
        });


        holder.btn_decrease.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(minteger!=0){
                    minteger = minteger - 1;
                    holder.tv_integer_number.setText(String.valueOf(minteger));
                    if(minteger==0){
                        holder.btn_buy.setVisibility(View.VISIBLE);
                        holder.linear_quantity.setVisibility(View.GONE);
                    }
                }
            }
        });

        holder.cv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClickListener.OnActionClickQuestion(v,position);
            }
        });
    }

    class ItemViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {


        TextView name,price,textview_price_discount,tv_integer_number,tv_product_unit;
        SquareImageView image;
        CardView cv;
        MaterialButton btn_buy,btn_increase,btn_decrease;
        Button button_discount;
        LinearLayout linear_quantity;
        ProgressBar progressBar;

        public ItemViewHolder(View view) {
            super(view);
            name   =  view.findViewById(R.id.tvProductName);
            price  = view.findViewById(R.id.tvProductPrice);
            image  = view.findViewById(R.id.ivProduct);
            cv     = view.findViewById(R.id.cvProduct);
            button_discount = view.findViewById(R.id.btnDiscount);
            btn_buy = view.findViewById(R.id.btnBuy);
            btn_increase = view.findViewById(R.id.btnIncrease);
            btn_decrease = view.findViewById(R.id.btnDecrease);
            tv_integer_number = view.findViewById(R.id.tvIntegerNumber);
            tv_product_unit = view.findViewById(R.id.tvProductUnit);
            progressBar = view.findViewById(R.id.progress_bar);

            textview_price_discount = view.findViewById(R.id.tvProductDiscount);
            linear_quantity = view.findViewById(R.id.llQuantity);
        }

        @Override
        public void onClick(View v) {

        }
    }
    public void actionQuestion(OnItemClickListener actionQuestion) {
        onItemClickListener = actionQuestion;
    }public interface OnItemClickListener {
        void OnActionClickQuestion(View view, int position);
    }

}

File: app/src/main/java/sembako/sayunara/android/ui/component/product/listproduct/adapter/ProductAdapter.kt
package sembako.sayunara.android.ui.component.product.listproduct.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.Context.WINDOW_SERVICE
import android.graphics.Paint
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.desmond.squarecamera.SquareImageView
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import de.hdodenhof.circleimageview.CircleImageView
import sembako.sayunara.android.R
import sembako.sayunara.android.ui.component.product.listproduct.model.Product
import sembako.sayunara.android.ui.util.HFRecyclerViewAdapter
import java.lang.StringBuilder
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols

open class ProductAdapter(var context: Context?, isHome: Boolean, isLike: Boolean, isBuy : Boolean,mOnClickListener: OnClickListener) : HFRecyclerViewAdapter<Product?, ProductAdapter.ItemViewHolder?>(context) {

    private lateinit var mOnClickListener: OnClickListener
    var total = 0
    private var isHome = false
    private var isLike = false
    private var isBuy  = false
    var price = 0

    override fun footerOnVisibleItem() {

    }
    override fun onCreateDataItemViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view: View = if (isHome) {
            LayoutInflater.from(parent.context).inflate(R.layout.list_item_product, parent, false)
        } else {
            LayoutInflater.from(parent.context).inflate(R.layout.list_item_product, parent, false)
        }
        return ItemViewHolder(view)
    }

    inner class ItemViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        var tvProductName: TextView      = view.findViewById(R.id.tvProductName)
        var tvProductPrice: TextView     = view.findViewById(R.id.tvProductPrice)
        var tvProductDiscount: TextView  = view.findViewById(R.id.tvProductDiscount)
        var tvProductUnit: TextView      = view.findViewById(R.id.tvProductUnit)
        var ivProductImage: ImageView    = view.findViewById(R.id.ivProduct)
        var cvProduct: CardView          = view.findViewById(R.id.cvProduct)
        var btnBuy: Button               = view.findViewById(R.id.btnBuy)
        var btnDiscount: Button          = view.findViewById(R.id.btnDiscount)
        var llinear : LinearLayout = view.findViewById(R.id.llProduct)


    }

    interface OnClickListener {
        fun onClickDetail(position: Int,product:Product)
    }


    init {
        this.isHome = isHome
        this.isLike = isLike
        this.isBuy  = isBuy
        this.mOnClickListener = mOnClickListener
    }

    private fun setName(str: String): String? {
        val strArray = str.split(" ").toTypedArray()
        val builder = StringBuilder()
        for (s in strArray) {
            val cap = s.substring(0, 1).toUpperCase() + s.substring(1)
            val stm: String = cap
            builder.append("$stm ")
        }
        return builder.toString()
    }

    @SuppressLint("SetTextI18n")
    override fun onBindDataItemViewHolder(holder: ItemViewHolder?, position: Int) {
        val product = data[position]

        // val user = userList[position]

        holder?.tvProductName?.text = setName(product?.detail?.name.toString())
       // holder?.tvUsername?.text = user?.profile?.username

        if(product?.detail?.images?.size!=0){
            val urlImage = product?.detail?.images?.get(0)
            setImage(urlImage.toString(),holder?.ivProductImage)
        }

        val harga = product?.detail?.price.toString().toDouble()
        val df = DecimalFormat.getCurrencyInstance() as DecimalFormat
        val dfs = DecimalFormatSymbols()
        dfs.currencySymbol = ""
        dfs.monetaryDecimalSeparator = ','
        dfs.groupingSeparator = '.'
        df.decimalFormatSymbols = dfs
        val k = df.format(harga)
        if (product?.detail?.discount == 0L) {
            holder?.btnDiscount?.visibility = View.GONE
            holder?.tvProductPrice?.text = "Rp$k".dropLast(3)
            holder?.tvProductUnit?.text = product.detail?.weight.toString() + " " + product.detail?.unit
        } else {
            holder?.btnDiscount?.visibility = View.VISIBLE
            holder?.tvProductDiscount?.visibility = View.VISIBLE
            holder?.btnDiscount?.text = product?.detail?.discount.toString() + " %"
            val price = product?.detail?.price.toString().toInt()
            val discount = product?.detail?.discount.toString().toInt()
            val total = discount * price / 100
            val total2 = price - total
            val amount = total2.toString().toDouble()
            holder?.tvProductDiscount?.paintFlags =
                holder?.tvProductDiscount?.paintFlags!! or Paint.STRIKE_THRU_TEXT_FLAG
            holder.tvProductDiscount.text =
                "Rp" + k + " / " + product?.detail?.weight + " " + product?.detail?.unit
            holder.tvProductUnit.text = product?.detail?.weight.toString() + " " + product?.detail?.unit
            val harga = df.format(amount).dropLast(3)
            holder.tvProductPrice.text = "Rp$harga /"
        }
        holder?.btnBuy?.setOnClickListener {
          //  mOnClickListener.onClickDetail(position,product)
        }

        holder?.cvProduct?.setOnClickListener {
           // mOnClickListener.onClickDetail(position,product)
        }

        setupLayout(holder?.llinear)

       /* if(isSeller()||isAdmin()||isSuperAdmin()){
            if(product.status?.draft==true){
                holder.ivDraftImage.visibility = View.VISIBLE
            }
        }

        if(isCustomer()){
            if(getUser()?.profile?.userId == product.userId){
                holder.btnBuy.visibility = View.GONE
            }else{
                holder.btnBuy.visibility = View.VISIBLE
            }
            holder.ivAction.visibility = View.GONE
        }else{
            holder.btnBuy.visibility = View.GONE
            holder.ivAction.visibility = View.VISIBLE
        }

        holder.ivAction.setOnClickListener {
            mOnClickListener.onActionClick(position,product)
        }

        if(like==true){
            holder.rlLike.visibility=View.VISIBLE
        }else{
            holder.rlLike.visibility=View.GONE
        }

        if(isSuperAdmin()||isAdmin()||isSeller()){
            holder.layoutStatus.visibility = View.VISIBLE
            if(product.status?.active==false){
                //Delete By Owner
                holder.tvStatus.text = "Status Deleted"
            }else{
                if(product.status?.publish==true){
                    holder.tvStatus.text = "Status Publish"
                }else{
                    holder.tvStatus.text = "Status UnPublish"
                }

                if(product.status?.draft==true){
                    holder.tvStatus.text = "Status Draft"
                }
            }
        }*/

       /* if(user?.profile?.avatar!=""){
            Picasso.get()
                .load(user?.profile?.avatar.toString())
                .into(holder?.ivUser)
        }*/
    }

    private fun setupLayout(linearLayout: LinearLayout?){
        val windowManager = context?.getSystemService(WINDOW_SERVICE) as WindowManager
        val displayMetrics = DisplayMetrics()
        val display = windowManager.defaultDisplay
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        val width = (display.width /2)-50//- adjustScreen;  // deprecated
        val height = display.height
        val params = linearLayout?.layoutParams //as LinearLayout.LayoutParams
        params?.width = width
        params?.height = height
        linearLayout?.layoutParams = params
    }

    private fun setImage(url : String, imageView : ImageView?){
        Picasso.get()
            .load(url)
            .into(imageView)

    }

}